pennylane/templates/state_preparations/superposition.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the Superposition template.
+"""
+
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+
+
+def _get_permutation(basis_list):
+    r"""
+    Given a list of :math:`m` basis states, this function generates a dictionary assigning to each of them
+    the :math:`m`-th basis states in the computational base. Also, if a state within ``basis_list`` is one
+    of the first :math:`m` basis states, this state will be assigned to itself.
+
+    ** Example **
+
+    .. code-block:: pycon
+
+        >>> basis_list = [[1, 1, 0, 0], [1, 0, 1, 0], [0, 1, 0, 1], [1, 0, 0, 1]]
+        >>> _get_permutation(basis_list)
+        {
+        [1, 1, 0, 0]: [0, 0, 0, 0],
+        [1, 0, 1, 0]: [0, 0, 0, 1],
+        [0, 1, 0, 1]: [0, 0, 1, 0],
+        [1, 0, 0, 1]: [0, 0, 1, 1]
+        }
+
+
+    .. code-block:: pycon
+
+        >>>> basis_list = [[1, 1, 0, 0], [0, 1, 0, 1], [0, 0, 0, 1], [1, 0, 0, 1]]
+        >>> _get_permutation(basis_list)
+        {
+        [1, 1, 0, 0]: [0, 0, 0, 0],
+        [0, 1, 0, 1]: [0, 0, 1, 0],
+        [0, 0, 0, 1]: [0, 0, 0, 1],
+        [1, 0, 0, 1]: [0, 0, 1, 1]
+        }
+
+    """
+
+    length = len(basis_list[0])
+    smallest_basis_lists = [tuple(map(int, f"{i:0{length}b}")) for i in range(len(basis_list))]
+
+    binary_dict = {}
+    used_smallest = set()
+
+    # Assign keys that can map to themselves
+    for original in basis_list:
+
+        if original in smallest_basis_lists and tuple(original) not in used_smallest:
+
+            binary_dict[tuple(original)] = original
+            used_smallest.add(tuple(original))
```

Comment:
So, if a state appears in `basis_list` multiple times the duplicates are just ignored? Maybe this should raise an exception because it doesn't make sense to put two copies of the same state in superposition.

Reply:
That's a good idea, I'll add a error if the basis provided is not unique!